src/components/UploadAvatarModal/UploadAvatarModal.js

```diff
@@ -0,0 +1,247 @@
+        const canvas = document.createElement('canvas')
+        const ctx = canvas.getContext('2d')
+
+        const maxSize = Math.max(image.width, image.height)
+        const safeArea = 2 * ((maxSize / 2) * Math.sqrt(2))
+
+        // set each dimensions to double largest dimension to allow for a safe area for the
+        // image to rotate in without being clipped by canvas context
+        canvas.width = safeArea
+        canvas.height = safeArea
+
+        // translate canvas context to a central location on image to allow rotating around the center.
+        ctx.translate(safeArea / 2, safeArea / 2)
+        ctx.rotate((rotation * Math.PI) / HALF_CIRCLE)
+        ctx.translate(-safeArea / 2, -safeArea / 2)
+
+        // draw rotated image and store data.
+        ctx.drawImage(
+          image,
+          safeArea / 2 - image.width / 2,
+          safeArea / 2 - image.height / 2,
+        )
+        const data = ctx.getImageData(0, 0, safeArea, safeArea)
+
+        // set canvas width to final desired crop size - this will clear existing context
+        canvas.width = croppedAreaPixels.width
+        canvas.height = croppedAreaPixels.height
+
+        // paste generated rotate image with correct offsets for x,y crop values.
+        ctx.putImageData(
+          data,
+          Math.round(0 - safeArea / 2 + image.width / 2 - croppedAreaPixels.x),
+          Math.round(0 - safeArea / 2 + image.height / 2 - croppedAreaPixels.y),
+        )
+
+        // Convert the contents of the canvas blob to an image
+        const reader = new FileReader()
+        canvas.toBlob((blob) => {
+          reader.readAsDataURL(blob)
+          reader.onloadend = () => {
+            const [, mime, b64data] = reader.result.match(/:([^;]+);[^,]+,(.*)$/u)
+            const bstr = atob(b64data)
+            let datalength = bstr.length
+            const u8arr = new Uint8Array(datalength)
+
+            while (datalength > 0) {
+              datalength -= 1
+              u8arr[datalength] = bstr.charCodeAt(datalength)
+            }
+            const croppedImage = new File([u8arr], 'avatar.png', { type: mime })
+
+            if (croppedImage.size > MAX_FILE_SIZE) {
+              setResult({
+                error: { status: 'toobig' },
+                success: false,
+                submitted: false,
+              })
+            } else {
+              submit(mime, croppedImage)
+            }
+          }
+        })
+      })
+      image.src = upImg
+    } catch (exception) {
+      console.error(exception)
+      setResult({
+        error: { status: 'internal' },
+        success: false,
+        submitted: false,
+      })
+    }
+  }, [croppedAreaPixels, rotation, submit, upImg])
+
+  return (
+    <ModalContent className="dialog no-pad">
+      <UploadAvatarMessageBox result={result} />
+      <div>
+        <label>{'Select Image: '}</label>
+        <input accept="image/*" aria-label="Select Image" type="file" onChange={onSelectFile} />
```

Comment:
Would be great if we could limit this to just the file types we accept. See ticket for supported image types. all of their MIMES are pretty straightforward.

Reply:
As far as I'm aware, all images ultimately uploaded to the server are *going* to be png, because of the cropper tool drawing whatever image to a hidden canvas for the transforms, then pulling the data of the canvas off (which comes off as png). Is there a need to restrict what the user can do in this case? Or should we just check that the canvas data being pulled off is one of the three types?

Reply:
Uhh that's a fair point. Could we test for GIFs to see how they get processed? I assume it'll just take the first frame but I wanna be sure.

Reply:
I'd not thought of animated gifs. Because the cropper is a standard <img> it'll animate it while editing, but when you click "upload" it does indeed only show the first frame. It gives the impression your cropped avatar will animate, which it won't. I think that's a good enough reason not to have image/* :(